contacts/views.py

```diff
@@ -0,0 +1,96 @@
+from django.db.utils import IntegrityError
+from rest_framework import generics, permissions, status
+from rest_framework.exceptions import ValidationError
+from rest_framework.response import Response
+
+from .models import Case, Contact
+from .serializers import CaseSerializer, ContactSerializer
+
+
+class ConfirmedContactView(generics.GenericAPIView):
+    """
+    POST endpoint to create a case and contact.
+    ---
+    POST fields:
+    * msisdn
+    * timestamp
+    * external_id
+    - name (optional)
+    - case_id (optional)
+    ---
+    Response example:
+    ```
+    {
+        "id": "84fe2a6c-a5f8-4994-9d96-d765514740b3",
+        "msisdn": "+27820001001",
+        "date_start": "2020-07-19T19:35:11.400070Z",
+        "case_id": null,
+        "name": null,
+        "created_by": "admin",
+        "created_at": "2020-07-21T13:09:06.921305Z",
+        "external_id": "51e7f725b59442c7b75b48752b29f243"
+    }
+    ```
+    """
+
+    permission_classes = (permissions.IsAuthenticated,)
+
+    def post(self, request, *args, **kwargs):
+        data = request.data
+
+        msisdn = data.pop("msisdn", None)
+
+        # Try to create a contact
+
+        contact_query = Contact.objects.filter(msisdn=msisdn)
+        if contact_query.exists():
+            contact = contact_query.first()
+        else:
+            try:
+                contact_serializer = ContactSerializer(data={"msisdn": msisdn})
+                contact_serializer.is_valid(raise_exception=True)
+                contact = contact_serializer.save()
+            except (IntegrityError, ValidationError):
```

Comment:
I think we should be handling `IntegrityError` separately to `ValidationError`.

`ValidationError` means that the MSISDN supplied is invalid, so we should return a 400 response.
`IntegrityError` means a race condition, so we should do a `Contact.objects.get(msisdn=msisnd)`, and not return an error to the user.

Reply:
For the `ValidationError`, if we don't catch it here, ie. just remove it from the list, then won't rest framework handle returning a 400 with a descriptive error for us?

Reply:
Done. For some reson i decided to catch validation error - that was wrong. Now invalid fields are raising 400 response correctly.